File: frontend/src/types/sse.ts
export const ANALYSIS_STATUSES = [
  "queued",
  "running",
  "partial",
  "completed",
  "failed",
] as const;

export const ANALYSIS_STEPS = [
  "upload",
  "transcribe",
  "speech",
  "content",
  "finalize",
] as const;

export const ANALYSIS_EVENT_NAMES = [
  "analysis.status",
  "analysis.partial",
  "analysis.result",
] as const;

export type AnalysisStatus = (typeof ANALYSIS_STATUSES)[number];
export type AnalysisStep = (typeof ANALYSIS_STEPS)[number];
export type AnalysisEventName = (typeof ANALYSIS_EVENT_NAMES)[number];

export type EventError = {
  code: string;
  detail: string;
};

export type BaseAnalysisEvent = {
  event: AnalysisEventName;
  id?: string;
  run_id: string;
  owner_sub: string;
  status: AnalysisStatus;
  step: AnalysisStep;
  emitted_at: string;
  message?: string;
  data?: Record<string, unknown>;
  error?: EventError;
};

export type AnalysisStatusEvent = BaseAnalysisEvent & {
  event: "analysis.status";
};

export type AnalysisPartialData = {
  partial_summary?: string;
  [key: string]: unknown;
};

export type AnalysisPartialEvent = BaseAnalysisEvent & {
  event: "analysis.partial";
  status: "partial";
  data?: AnalysisPartialData;
};

export type AnalysisResultData = {
  summary: string;
  strengths: string[];
  improvements: string[];
  agent_cost: AgentExecutionCost;
  file_name?: string;
  file_path?: string;
  [key: string]: unknown;
};

export type AgentExecutionCostBreakdown = {
  orchestrator_usd?: number;
  speech_analyzer_usd?: number;
  content_analyzer_usd?: number;
  [key: string]: number | undefined;
};

export type AgentExecutionCost = {
  total_usd: number;
  breakdown: AgentExecutionCostBreakdown;
  is_estimated: boolean;
  pricing_version: string;
};

export type AnalysisResultEvent = BaseAnalysisEvent & {
  event: "analysis.result";
  status: "completed";
  step: "finalize";
  data: AnalysisResultData;
};

export type AnalysisEvent =
  | AnalysisStatusEvent
  | AnalysisPartialEvent
  | AnalysisResultEvent;

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null;
}

function isNumber(value: unknown): value is number {
  return typeof value === "number" && Number.isFinite(value);
}

function isEnumValue<T extends readonly string[]>(
  source: T,
  value: unknown,
): value is T[number] {
  return typeof value === "string" && source.includes(value);
}

export function isAnalysisEvent(value: unknown): value is AnalysisEvent {
  if (!isRecord(value)) return false;

  if (!isEnumValue(ANALYSIS_EVENT_NAMES, value.event)) return false;
  if (!isEnumValue(ANALYSIS_STATUSES, value.status)) return false;
  if (!isEnumValue(ANALYSIS_STEPS, value.step)) return false;
  if (typeof value.run_id !== "string" || value.run_id.length === 0) return false;
  if (typeof value.owner_sub !== "string" || value.owner_sub.length === 0)
    return false;
  if (typeof value.emitted_at !== "string" || value.emitted_at.length === 0)
    return false;

  if (value.id !== undefined && typeof value.id !== "string") return false;
  if (value.message !== undefined && typeof value.message !== "string")
    return false;

  if (value.error !== undefined) {
    if (!isRecord(value.error)) return false;
    if (typeof value.error.code !== "string") return false;
    if (typeof value.error.detail !== "string") return false;
  }

  if (value.event === "analysis.partial") {
    if (value.status !== "partial") return false;
    if (value.data !== undefined) {
      if (!isRecord(value.data)) return false;
      if (
        value.data.partial_summary !== undefined &&
        typeof value.data.partial_summary !== "string"
      )
        return false;
    }
  }

  if (value.event === "analysis.result") {
    if (value.status !== "completed" || value.step !== "finalize") return false;
    if (!isRecord(value.data)) return false;
    if (typeof value.data.summary !== "string") return false;
    if (
      !Array.isArray(value.data.strengths) ||
      !value.data.strengths.every((item) => typeof item === "string")
    ) {
      return false;
    }
    if (
      !Array.isArray(value.data.improvements) ||
      !value.data.improvements.every((item) => typeof item === "string")
    ) {
      return false;
    }

    if (!isRecord(value.data.agent_cost)) return false;
    if (!isNumber(value.data.agent_cost.total_usd)) return false;
    if (!isRecord(value.data.agent_cost.breakdown)) return false;
    if (
      Object.values(value.data.agent_cost.breakdown).some(
        (item) => item !== undefined && !isNumber(item),
      )
    ) {
      return false;
    }
    if (typeof value.data.agent_cost.is_estimated !== "boolean") return false;
    if (typeof value.data.agent_cost.pricing_version !== "string") return false;
  }

  return true;
}

File: frontend/amplify/agent/runtime/events/sse.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timezone
from enum import StrEnum
import json
from typing import Any, Mapping
from uuid import uuid4


class AnalysisStatus(StrEnum):
    QUEUED = "queued"
    RUNNING = "running"
    PARTIAL = "partial"
    COMPLETED = "completed"
    FAILED = "failed"


class AnalysisStep(StrEnum):
    UPLOAD = "upload"
    TRANSCRIBE = "transcribe"
    SPEECH = "speech"
    CONTENT = "content"
    FINALIZE = "finalize"


class AnalysisEventName(StrEnum):
    STATUS = "analysis.status"
    PARTIAL = "analysis.partial"
    RESULT = "analysis.result"


@dataclass(slots=True)
class EventError:
    code: str
    detail: str

    def to_dict(self) -> dict[str, str]:
        return {"code": self.code, "detail": self.detail}


@dataclass(slots=True)
class AgentExecutionCostSummary:
    total_usd: float
    breakdown: dict[str, float]
    is_estimated: bool = True
    pricing_version: str = "unknown"

    def to_dict(self) -> dict[str, Any]:
        if self.total_usd < 0:
            raise ValueError("total_usd must be >= 0")
        if any(value < 0 for value in self.breakdown.values()):
            raise ValueError("breakdown values must be >= 0")

        return {
            "total_usd": round(self.total_usd, 6),
            "breakdown": {key: round(value, 6) for key, value in self.breakdown.items()},
            "is_estimated": self.is_estimated,
            "pricing_version": self.pricing_version,
        }


def utc_now_iso8601() -> str:
    return datetime.now(timezone.utc).replace(microsecond=0).isoformat().replace("+00:00", "Z")


def new_analysis_event(
    *,
    event: AnalysisEventName,
    run_id: str,
    owner_sub: str,
    status: AnalysisStatus,
    step: AnalysisStep,
    message: str | None = None,
    data: dict[str, Any] | None = None,
    error: EventError | None = None,
    event_id: str | None = None,
    emitted_at: str | None = None,
) -> dict[str, Any]:
    payload: dict[str, Any] = {
        "event": event.value,
        "id": event_id or f"evt_{uuid4()}",
        "run_id": run_id,
        "owner_sub": owner_sub,
        "status": status.value,
        "step": step.value,
        "emitted_at": emitted_at or utc_now_iso8601(),
    }
    if message:
        payload["message"] = message
    if data is not None:
        payload["data"] = data
    if error is not None:
        payload["error"] = error.to_dict()

    return payload


def new_analysis_result_event(
    *,
    run_id: str,
    owner_sub: str,
    summary: str,
    strengths: list[dict] | list[str],
    improvements: list[dict] | list[str],
    detailed_feedback: str = "",
    agent_cost: AgentExecutionCostSummary,
    file_name: str | None = None,
    file_path: str | None = None,
    transcript: str | None = None,
    transcript_s3_key: str | None = None,
    message: str | None = None,
    event_id: str | None = None,
    emitted_at: str | None = None,
) -> dict[str, Any]:
    data: dict[str, Any] = {
        "summary": summary,
        "strengths": strengths,
        "improvements": improvements,
        "detailed_feedback": detailed_feedback,
        "agent_cost": agent_cost.to_dict(),
    }
    if file_name is not None:
        data["file_name"] = file_name
    if file_path is not None:
        data["file_path"] = file_path
    if transcript is not None:
        data["transcript"] = transcript
    if transcript_s3_key is not None:
        data["transcript_s3_key"] = transcript_s3_key
    return new_analysis_event(
        event=AnalysisEventName.RESULT,
        run_id=run_id,
        owner_sub=owner_sub,
        status=AnalysisStatus.COMPLETED,
        step=AnalysisStep.FINALIZE,
        message=message,
        data=data,
        event_id=event_id,
        emitted_at=emitted_at,
    )


def format_sse_message(payload: Mapping[str, Any]) -> str:
    event_name = payload.get("event")
    if not isinstance(event_name, str) or not event_name:
        raise ValueError("payload.event is required")

    lines: list[str] = []

    event_id = payload.get("id")
    if isinstance(event_id, str) and event_id:
        lines.append(f"id: {event_id}")

    lines.append(f"event: {event_name}")
    lines.append(f"data: {json.dumps(dict(payload), ensure_ascii=False)}")
    lines.append("")
    return "\n".join(lines)


def format_error_event(
    *,
    run_id: str,
    owner_sub: str,
    step: AnalysisStep,
    code: str,
    detail: str,
    event_id: str | None = None,
) -> str:
    payload = new_analysis_event(
        event=AnalysisEventName.STATUS,
        run_id=run_id,
        owner_sub=owner_sub,
        status=AnalysisStatus.FAILED,
        step=step,
        error=EventError(code=code, detail=detail),
        event_id=event_id,
    )
    return format_sse_message(payload)


def assert_event_ownership(*, request_sub: str, owner_sub: str) -> None:
    if request_sub != owner_sub:
        raise PermissionError("forbidden: run_id does not belong to request user")
